linkerd/app/outbound/src/http/logical/policy.rs

```diff
@@ -0,0 +1,396 @@
+                // `SelectRoute` impl.
+                .push_on_service(RouteParams::layer())
+                .push(svc::NewOneshotRoute::<Self, (), _>::layer_cached())
+                .push(svc::ArcNewService::layer())
+                .into_inner()
+        })
+    }
+}
+
+impl<T, M, F> From<(RouterRoutes<M, F>, T)> for RouterParams<T, M, F>
+where
+    T: Eq + Hash + Clone + Debug,
+    M: Clone,
+    F: Clone,
+{
+    fn from((rts, parent): (RouterRoutes<M, F>, T)) -> Self {
+        let RouterRoutes {
+            addr,
+            routes,
+            backends,
+        } = rts;
+
+        let mk_concrete = {
+            let authority = addr.to_http_authority();
+            let parent = parent.clone();
+            move |target: concrete::Dispatch| Concrete {
+                target,
+                authority: Some(authority.clone()),
+                parent: parent.clone(),
+            }
+        };
+
+        let mk_dispatch = move |bd: &policy::BackendDispatcher| match *bd {
+            policy::BackendDispatcher::BalanceP2c(
+                policy::Load::PeakEwma(policy::PeakEwma { decay, default_rtt }),
+                policy::EndpointDiscovery::DestinationGet { ref path },
+            ) => mk_concrete(concrete::Dispatch::Balance(
+                path.parse().expect("destination must be a nameaddr"),
+                balance::EwmaConfig { decay, default_rtt },
+            )),
+            policy::BackendDispatcher::Forward(addr, ref metadata) => mk_concrete(
+                concrete::Dispatch::Forward(Remote(ServerAddr(addr)), metadata.clone()),
+            ),
+        };
+
+        let mk_distribution = |d: &policy::RouteDistribution<F>| match d {
+            policy::RouteDistribution::Empty => Distribution::Empty,
+            policy::RouteDistribution::FirstAvailable(backends) => Distribution::first_available(
+                backends
+                    .iter()
+                    .map(|rb| mk_dispatch(&rb.backend.dispatcher)),
+            ),
+            policy::RouteDistribution::RandomAvailable(backends) => Distribution::random_available(
+                backends
+                    .iter()
+                    .map(|(rb, weight)| (mk_dispatch(&rb.backend.dispatcher), *weight)),
+            )
+            .expect("distribution must be valid"),
+        };
+
+        let mk_policy = |policy::RoutePolicy::<F> {
+                             meta,
+                             filters,
+                             distribution,
+                         }| RouteParams {
+            addr: addr.clone(),
+            parent: parent.clone(),
+            meta,
+            filters,
+            distribution: mk_distribution(&distribution),
+        };
+
+        let routes = routes
+            .iter()
+            .map(|route| route::Route {
+                hosts: route.hosts.clone(),
+                rules: route
+                    .rules
+                    .iter()
+                    .cloned()
```

Comment:
almost left a comment that this clones `distribution` unnecessarily: it's only `meta` and `filters` that are cloned into the `RouteParams`, so we could iterate by reference and have `mk_policy` take an `&RoutePolicy`, and clone just `meta` and `filters`. but, the distribution is also `arc`ed so this doesn't actually matter that much...